products/memgraph/upcoming/tutorials/understanding-music-with-modules.md

```diff
@@ -0,0 +1,450 @@
+
+When using Memgraph installed from DEB or RPM package, the currently running
+Memgraph server may need to be stopped before importing the example. The user
+can do so using the following command:
+
+```plaintext
+systemctl stop memgraph
+```
+
+When using Docker, the example can be imported with the following command
+(notice how we need to mount the query module directory first):
+
+```plaintext
+docker run -p 7687:7687 \
+  -v mg_lib:/var/lib/memgraph -v mg_log:/var/log/memgraph -v mg_etc:/etc/memgraph \
+  -v $(pwd)/path/to/your/modules/directory:/path/to/mount/of/modules/directory \
+  memgraph --data-directory /usr/share/memgraph/examples/Europe \
+  --storage-snapshot-interval-sec=0 --storage-wal-enabled=false \
+  --storage-snapshot-on-exit=false --storage-properties-on-edges=true
+  --query-modules-directory=/path/to/mount/of/modules/directory
+```
+
+The user should note that any modifications of the database state will persist
+only during this run of Memgraph.
+
+### Example Queries and procedures
+
+First, let's create a Python script in our modules directory which will contain
+our procedures. We'll name the script `deezer_example.py`.
+
+After each change to the script, we need to tell Memgraph to reload all the
+modules. We can do that by calling the following command:
+
+```opencypher
+CALL mg.load_all();
+```
+
+Let's count how many times a specific genre occurs in our network!
+
+We will define a procedure called `genre_count`:
+
+```python
+import mgp
+
+@mgp.read_proc
+def genre_count(context: mgp.ProcCtx,
+                genre: str
+               ) -> mgp.Record(genre=str, count=int):
+    count = len(
+        [v for v in context.graph.vertices if genre in v.properties['genres']]
+    )
+    return mgp.Record(genre=genre, count=count)
+```
+
+Let's reload the modules and check the results:
+
+```opencypher
+CALL deezer_example.genre_count("Pop")
+YIELD genre, count
+RETURN genre, count;
+```
+
+We can notice multiple things:
+
+- import of the `mgp` module which contains helper functions and types for
+  defining custom procedures
+- `@mgp.read_proc` which marks the function as a procedure
+- every argument is defined with a type
+- result is defined as an object of `mgp.Record` which also has defined types of
+  its members
+
+This example is probably not that interesting to you because we can get the same
+result using the following query:
+
+```opencypher
+MATCH (n)
+WITH n, "Pop" as genre
+WHERE genre in n.genres
+RETURN genre, count(n);
+```
+
+Let's try something more interesting. The genres are listed in the order the
+users have added them. If we assume that users picked the genres in order of
+preference, let's write a function that tells us in what percentage each genre
+appears in top n places.
+
+```python
+from collections import defaultdict
+
+@mgp.read_proc
+def in_top_n_percentage(context: mgp.ProcCtx,
+                        n: int
+                       ) -> mgp.Record(in_top_n_percentages=list):
```

Comment:
You can use the generic `mgp.List` type annotation if you know that the list is homogeneous (i.e. all of the elements are of the same type). In this case you could use `mgp.List[mgp.Map]` which translates to Cypher's `LIST OF MAP` type.

`list` is useful for heterogeneous lists since it translates to Cypher's `LIST OF ANY?` type (see the [Cypher typing proposal](https://github.com/opencypher/openCypher/blob/master/cip/1.accepted/CIP2015-09-16-public-type-system-type-annotation.adoc) for details about the type system).